### PR TITLE
Nuke: Build template from workfile

### PR DESCRIPTION
**Pull request by jrsndl**
## Brief description
Following discussion https://github.com/pypeclub/OpenPype/discussions/2013
Similar to PR https://github.com/pypeclub/OpenPype/pull/3544, but allows making template from workfile in one go, and works with creators. 

There is already a mechanism for loading Nuke template scripts, even automatically on a first open. Nuke workfile Builder settings allow to create first workfile and run Builder on first launch. This PR allows to create template from workfile with one click. It connects loaders and creators, and applies (some) knob values from the template, for example  publishing and Deadline settings.

## Description
For propagating master shots, prepare the master shot, and when happy, convert it to the template Nuke script.
Save the template Nuke script to the location set up in the workfile builder settings, for example ```{root[work]}/{project[name]}/templates/{task[name]}.nk```. When artists launches the task with matching task name, your master script will autoload as version 1, complete with correct loaders and creators.

## Simple Template 
Showing two Loaders and one Write Render
image

## Shot with Template Applied
from above, at first open
image

## Tested Workfile Builder settings
Settings in json:
[WorkfileBuilder.json.txt](https://github.com/pypeclub/OpenPype/files/9362997/WorkfileBuilder.json.txt)

image

## Testing notes:
1. Set up the Workfile Builder settings
2. Work on first Nuke task as usual. Save
3. Run ```OpenPype - Workfile to Template```. Note that if you select some nodes, only selection will be templated
4. Use Save as to save the template to the location specified by settings workfile_builder - curstom_templates - path
5. Launch Nuke in context of other task. If no version exists, template will auto apply
6. Alternatively, you can run ```OpenPype - Builde Workfile and Apply Template``` from the top menu

### CODE REVIEW

**Comment by jrsndl** [comment]
> 
I will check PR https://github.com/pypeclub/OpenPype/pull/3578. Thx.
I aimed on ease of use and self building first version. I have similar stuff in production (OP2) for long time, can't be without it...

**Review by jakubjezek001 — CHANGES_REQUESTED** [review]
I recommend to enclose the Settings which should be set as a Jason file into testing notes.

I hope all is set according your expectations. But it is not working yet. Here is Traceback:

```
Traceback (most recent call last):
  File "C:/Program Files/Nuke13.0v1/plugins\nuke_internal\callbacks.py", line 72, in onCreate
    _doCallbacks(onCreates)
  File "C:/Program Files/Nuke13.0v1/plugins\nuke_internal\callbacks.py", line 46, in _doCallbacks
    f[0](*f[1],**f[2])
  File "C:\CODE\__PYPE\OpenPype\openpype\hosts\nuke\api\lib.py", line 2523, in process_workfile_builder
    apply_template_nodes(run_build=False)
  File "C:\CODE\__PYPE\OpenPype\openpype\hosts\nuke\api\lib.py", line 2486, in apply_template_nodes
    template.apply_template()
  File "C:\CODE\__PYPE\OpenPype\openpype\hosts\nuke\api\workfile_template.py", line 337, in apply_template
    self._template_connect()
  File "C:\CODE\__PYPE\OpenPype\openpype\hosts\nuke\api\workfile_template.py", line 301, in _template_connect
    knob_value = int(knob_value)
ValueError: invalid literal for int() with base 10: '50.0'
```

File: openpype/hosts/nuke/api/workfile_template.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+                    except NameError:
+                        pass
+
+            if target_node:
+                self._swap_nodes(one_template, target_node)
+                for knob_name, knob in knob_dict.items():
+                    try:
+                        knob_value = str(knob['value'])
+                        knob_type = int(knob['type'])
+                        if knob_type == 3:
```
Comment (jakubjezek001) [inline openpype/hosts/nuke/api/workfile_template.py:300]: If knob types were aligned with Setting's style you could rather use `openpype\hosts\nuke\api\lib.py:set_node_knobs_from_settings` for setting knob values.

**Comment by jakubjezek001** [comment]
But in overall I really like the approach, great work @jrsndl ! I cannot wait to introduce this workflow in productions ;)

**Comment by jrsndl** [comment]
I fixed the type bug for Nuke 13, and updated the top description with settings in json

**Comment by ClementHector** [comment]
@jrsndl What features are you missing in #3544?

**Comment by jrsndl** [comment]
> @jrsndl What features are you missing in #3544?

## Ease of use
I do miss an easy way to create template from working Nuke script. For artists or io guys prepping the shots, create placeholder dialog is maybe too technical.

## More Loaders
The template for Nuke should be able to handle not only Loadclip and Nukenode loaders, but also at least the LoadEffectsInputProcess often used for look.

## Creators
Having working template for creators is also important, especially for shots with multiple outputs. At least WriteRender and WritePrerender.

## Joining template with WorkfileBuilder
WorkfileBuilder already present in OpenPype allows for nice template filtering for task type and task name. So you can have a template for roto, cleanup and comp task names. Another advantage of joining template with WorkfileBuilder is already present feature of auto creating first version of the work file on opening.

The end goal is to have first version of the workfile generated by template and published for editorial check, all automatically.

**Comment by ClementHector** [comment]
image
This template is generated by user or code?

**Comment by jrsndl** [comment]
@ClementHector sorry for the late reply.

> image
> Is this template node user generated or code generated?

This template is code generated from Nuke script. Basically, OP loaders are "converted" to the template nodes.

I see now that your approach with regex filters is much more flexible. Maybe, to keep gest from both ways, the solution might be to make a Nuke script that "converts" from already existing Nuke loaders to your template nodes. This would keep the great flexibility, and get the easy start for artists as well?

### Creators
For Nuke, creators are "asset specific", so "generic" creators do not work for Nuke. Having creators (for example the main output) ready from the template is much appreciated by junior artists. It would be awesome to have it.

> Design for all software
> The code was planned for a simpler integration in all software, we even made a template builder for photoshop on the same code abstraction as maya and nuke.

This is the reason why your [PR](https://github.com/pypeclub/OpenPype/pull/3544) is so much better then my attempt. Thank you for the awesome work!

**Comment by jakubjezek001** [comment]
Closing as it is replaced by https://github.com/pypeclub/OpenPype/pull/3838

**Pull request closed without merging**